### PR TITLE
Fix incorrect `window_end` calculation for ReduceWindowOp

### LINKED ISSUE #1662: Possible bug in Reduce/ReduceWindow
### What happened?

The following test has different outputs for the StableHLO interpreter vs JAX eval:
```
// RUN: stablehlo-translate --interpret %s

module @jit__lambda_ attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main() {
    %0 = stablehlo.constant dense<[[0.477156043, 0.865567684, 0.613974452, 0.223835945, 0.887297034, 0.918688178, 0.372403026, 0.864771842, 0.207958102, 0.616077662, 0.91525054, 0.292452455, 0.13126111, 0.356621504, 0.537083149, 0.221684456], [0.577118754, 0.736410499, 0.296871781, 0.238048315, 0.694297552, 0.355420589, 0.473695874, 0.449064136, 0.873826861, 0.824459075, 0.174015164, 0.713999152, 0.170108438, 0.883243799, 0.594961166, 0.548542619], [0.100502849, 0.688369036, 0.494347215, 3.108610e-01, 0.121291637, 0.406444669, 0.822417378, 0.890864611, 0.0846705436, 0.524221301, 0.243774414, 0.385059476, 0.985714197, 0.59201169, 0.473546147, 0.479912877], [7.079710e-01, 0.752919555, 0.468122482, 0.110772491, 0.519748211, 0.263927698, 0.0270278454, 0.744363666, 0.403088212, 0.0132240057, 0.520995498, 0.0692948103, 0.0997607707, 0.804367661, 0.669661164, 0.848182678]]> : tensor<4x16xf32>
    %1 = stablehlo.constant dense<0.000000e+00> : tensor<6x8x6x2x4xf32>
    %2 = stablehlo.constant dense<"0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000072F6C743000000000000000072F6C74372F6C743000000000000000072F6C7434785CE4300000000000000004785CE434785CE4300000000000000004785CE4372F6C743000000000000000072F6C74372F6C743000000000000000072F6C74300000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000006FCD743000000000000000006FCD74306FCD743000000000000000006FCD7435E11DF4300000000000000005E11DF435E11DF4300000000000000005E11DF4306FCD743000000000000000006FCD74306FCD743000000000000000006FCD74300000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000072F6C743000000000000000072F6C74372F6C743000000000000000072F6C7434785CE4300000000000000004785CE434785CE4300000000000000004785CE4372F6C743000000000000000072F6C74372F6C743000000000000000072F6C7430000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003E8AB64300000000000000003E8AB6433E8AB64300000000000000003E8AB643CC86BC430000000000000000CC86BC43CC86BC430000000000000000CC86BC433E8AB64300000000000000003E8AB6433E8AB64300000000000000003E8AB643000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000C6FFA8430000000000000000C6FFA843C6FFA8430000000000000000C6FFA843A58AAE430000000000000000A58AAE43A58AAE430000000000000000A58AAE43C6FFA8430000000000000000C6FFA843C6FFA8430000000000000000C6FFA8430000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003E8AB64300000000000000003E8AB6433E8AB64300000000000000003E8AB643CC86BC430000000000000000CC86BC43CC86BC430000000000000000CC86BC433E8AB64300000000000000003E8AB6433E8AB64300000000000000003E8AB643000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000C6FFA8430000000000000000C6FFA843C6FFA8430000000000000000C6FFA843A58AAE430000000000000000A58AAE43A58AAE430000000000000000A58AAE43C6FFA8430000000000000000C6FFA843C6FFA8430000000000000000C6FFA84300000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000035469A43000000000000000035469A4335469A43000000000000000035469A4374559F43000000000000000074559F4374559F43000000000000000074559F4335469A43000000000000000035469A4335469A43000000000000000035469A4300000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000072F6C743000000000000000072F6C74372F6C743000000000000000072F6C7434785CE4300000000000000004785CE434785CE4300000000000000004785CE4372F6C743000000000000000072F6C74372F6C743000000000000000072F6C74300000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000006FCD743000000000000000006FCD74306FCD743000000000000000006FCD7435E11DF4300000000000000005E11DF435E11DF4300000000000000005E11DF4306FCD743000000000000000006FCD74306FCD743000000000000000006FCD74300000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000072F6C743000000000000000072F6C74372F6C743000000000000000072F6C7434785CE4300000000000000004785CE434785CE4300000000000000004785CE4372F6C743000000000000000072F6C74372F6C743000000000000000072F6C7430000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003E8AB64300000000000000003E8AB6433E8AB64300000000000000003E8AB643CC86BC430000000000000000CC86BC43CC86BC430000000000000000CC86BC433E8AB64300000000000000003E8AB6433E8AB64300000000000000003E8AB643000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000C6FFA8430000000000000000C6FFA843C6FFA8430000000000000000C6FFA843A58AAE430000000000000000A58AAE43A58AAE430000000000000000A58AAE43C6FFA8430000000000000000C6FFA843C6FFA8430000000000000000C6FFA8430000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003E8AB64300000000000000003E8AB6433E8AB64300000000000000003E8AB643CC86BC430000000000000000CC86BC43CC86BC430000000000000000CC86BC433E8AB64300000000000000003E8AB6433E8AB64300000000000000003E8AB643000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000C6FFA8430000000000000000C6FFA843C6FFA8430000000000000000C6FFA843A58AAE430000000000000000A58AAE43A58AAE430000000000000000A58AAE43C6FFA8430000000000000000C6FFA843C6FFA8430000000000000000C6FFA84300000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000035469A43000000000000000035469A4335469A43000000000000000035469A4374559F43000000000000000074559F4374559F43000000000000000074559F4335469A43000000000000000035469A4335469A43000000000000000035469A43000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"> : tensor<6x8x6x2x4xf32>
    %3 = stablehlo.broadcast_in_dim %0, dims = [3, 4] : (tensor<4x16xf32>) -> tensor<7x7x16x4x16xf32>
    %4 = stablehlo.constant dense<0.000000e+00> : tensor<f32>
    %5 = stablehlo.reduce(%3 init: %4) applies stablehlo.add across dimensions = [3, 4] : (tensor<7x7x16x4x16xf32>, tensor<f32>) -> tensor<7x7x16xf32>
    %6 = stablehlo.broadcast_in_dim %5, dims = [2, 3, 4] : (tensor<7x7x16xf32>) -> tensor<2x5x7x7x16xf32>
    %7 = stablehlo.reshape %6 : (tensor<2x5x7x7x16xf32>) -> tensor<7x7x16x2x5xf32>
    %8 = stablehlo.constant dense<0.000000e+00> : tensor<f32>
    %9 = stablehlo.broadcast_in_dim %8, dims = [] : (tensor<f32>) -> tensor<f32>
    %10 = "stablehlo.reduce_window"(%7, %9) ({
    ^bb0(%arg0: tensor<f32>, %arg1: tensor<f32>):
      %17 = stablehlo.add %arg0, %arg1 : tensor<f32>
      stablehlo.return %17 : tensor<f32>
    }) {base_dilations = dense<[3, 2, 4, 1, 3]> : tensor<5xi64>, padding = dense<[[4, 6], [2, 5], [6, 7], [6, 5], [3, 3]]> : tensor<5x2xi64>, window_dilations = dense<[3, 2, 2, 4, 3]> : tensor<5xi64>, window_dimensions = dense<[7, 7, 32, 3, 5]> : tensor<5xi64>, window_strides = dense<[2, 1, 2, 1, 2]> : tensor<5xi64>} : (tensor<7x7x16x2x5xf32>, tensor<f32>) -> tensor<6x8x6x5x4xf32>
    %11 = stablehlo.constant dense<0.000000e+00> : tensor<f32>
    %12 = stablehlo.reduce(%10 init: %11) applies stablehlo.add across dimensions = [3] : (tensor<6x8x6x5x4xf32>, tensor<f32>) -> tensor<6x8x6x4xf32>
    %13 = stablehlo.broadcast_in_dim %12, dims = [1, 2, 3, 4] : (tensor<6x8x6x4xf32>) -> tensor<2x6x8x6x4xf32>
    %14 = stablehlo.reshape %13 : (tensor<2x6x8x6x4xf32>) -> tensor<6x8x6x2x4xf32>
    %15 = stablehlo.maximum %14, %1 : tensor<6x8x6x2x4xf32>
    %16 = stablehlo.sqrt %15 : tensor<6x8x6x2x4xf32>
    check.expect_almost_eq %16, %2  : tensor<6x8x6x2x4xf32>
    func.return
  }
}
```

The error:
```
LLVM ERROR: Error evaluating function: main. 
	Fallback for check.expect_almost_eq failed: Element values don't match: 89.4252471 : f32 (actual) vs 399.925354 : f32 (expected) at index [1, 0, 0, 0, 0]
```

(The test was generated by lowering a JAX program to StableHLO. The "expected output" (`%2`) was calculated by evaluating the same JAX program on the inputs `%0` and `%1`.)

### PR DESCRIPTION
**Pull request by ghpvnist**
Turns out https://github.com/openxla/stablehlo/pull/1492 didn't actually fix the bug since `window_end` is still calculated incorrectly.

Shoutout to @pravnar for filing the bug :)

closes #1662

### CODE REVIEW

File: docs/spec.md [hunk 1]
```diff
@@ -4184,7 +4184,7 @@ where:
 * `padded_inputs = pad(inputs..., init_values..., padding[:, 0], padding[:, 1],
   base_dilations - 1)`.
 * `window_start = result_index * window_strides`.
-* `window_end = window_start + window_dimensions + window_dilations - 1`.
+* `window_end = window_start + window_dimensions + (window_dimensions - 1) * (window_dilations - 1)`.
```
Comment (burmako) [inline docs/spec.md:4187]: Let's use the formula from C15: `dilated_window_shape = window_dimensions = 0 ? 0 : (window_dimensions - 1) * window_dilations + 1`. It does the same thing, but also works correctly for the case when window_dimensions contains zeros.

Comment (burmako) [inline docs/spec.md:4187]: Also, let's add tests - both for the bug that this PR fixes and for the zero case.

**Review by burmako — CHANGES_REQUESTED** [review]

File: docs/spec.md (hunk 1 above)
Comment (ghpvnist) [reply docs/spec.md:4187]: To conform to SliceOp's `C3: 0 <= start_indices <= limit_indices <= shape(operand)`, let's use:
```
windowEnd = any_of(windowDimensions, isZero) ? windowStart : windowStart + (windowDimensions - 1) * windowDilations + 1;
```

Comment (burmako) [reply docs/spec.md:4187]: Sounds good. This makes sense for the interpreter - since even if a single dimension size is zero, then we don't need to do any computation. For shape inference, we'd need to do that on a per-dimension basis (so C15 is written the way it is).

Comment (ghpvnist) [reply docs/spec.md:4187]: Nvm. SliceOp's `C5: 0 < window_dimensions` doesn't allow `windowDimensions` to be 0 in the first place. Assuming we are using `stablehlo.slice`, this check is redundant. I think `windowEnd = windowStart + (windowDimensions - 1) * windowDilations + 1` is enough.

Comment (ghpvnist) [reply docs/spec.md]: Added a case for the bug, but not the zero case as noted in my comment above.

Comment (burmako) [reply docs/spec.md]: Good point. Let's update C15 as well then to drop the "window_dimensions = 0 ? 0 :" part there as well.

**Review by burmako — APPROVED** [review]
Minor comment called out above, and we're good to go.

File: docs/spec.md (hunk 1 above)
Comment (ghpvnist) [reply docs/spec.md]: Done.
